### DISCUSSION TITLE
Developing Go-Client

### DISCUSSION THREAD

**Discussion by wyyolo** [apache/streampipes#2398] [Ideas]
Many IoT companies are using Go language to develop projects. Streampipes already have Python and Java clients. Will the community consider developing a Go Client recently?

**Comment by luoluoyuyu** [comment]
Hi @wy-os 
Yes, many Internet companies in China are using Golang language for software development. I think adding go client can help more users.

**Reply by wyyolo** [reply]
Thank you for agreeing with my point of view. If the community has plans to develop Go-Client, I can provide assistance for it.

**Comment by tenthe** [comment]
Hi @wy-os,
thank you very much for starting the discussion. There are currently no plans to develop a Go client, as there are currently no Go developers in the community. However, this would certainly be a cool and useful extension. If you are interested in this topic, it would be great if you could provide support for it. 

I could imagine an integration similar to the Python client [1][2].

What do you think?

[1] https://github.com/apache/streampipes/tree/dev/streampipes-client-python
[2] https://streampipes.apache.org/docs/docs/python/latest/

Cheers,
Philipp

**Reply by wyyolo** [reply]
I'm glad to hear back from you . I think expanding multiple clients is beneficial for helping and attracting more developers to use StreamPipes. I believe that for StreamPipes users who know the Go language, it would be a great pleasure to participate in building Go-Client together.I think your viewpoint is worth adopting, and I will present a demo similar to Python-Client.

**Comment by RobertIndie** [comment]
+1 for having the Go client. This would also be very helpful to integrate StreamPipes with the K8S ecosystem.

**Reply by wyyolo** [reply]
I agree with your point of view and hope to have more communication with you.

**Comment by dominikriemer** [comment]
Hi @wy-os!
welcome and thank you for suggesting the feature! I also think that having Go support for StreamPipes would be a great addition.

Which feature set do you think is important for the Go Client? The Python client currently supports execution of functions and fetching data from the API, while Java also includes the full SDK for writing and running adapters and pipeline elements.

One thing where I think Go would be a very good choice is the ability to create lightweight adapters running on edge nodes, where Java adds too much overhead on resource-constrained devices. But this would require to also implement the SDK for adapters and pipeline elements in Go, which will probably be some effort.

So it would be great to discuss the targeted feature set and see who would be willing to contribute to the implementation.

**Reply by wyyolo** [reply]
Thank you very much for your reply !  I think that in the initial stage, we can implement a function similar to Python Client，and I also think it is necessary to support a complete SDK like Java to write and run adapters and pipeline elements.

I hope everyone can participate in the discussion of the target features of Go Client and work together to enrich StreamPipes. 

Lastly, thank you very much.

**Reply by wyyolo** [reply]
Also, as you mentioned, the idea of creating lightweight adapters to run on edge nodes is great!
